### PR TITLE
fix: [DHIS2-7986] Use Joda time to manipulate dates in the d2 functions

### PR DESCRIPTION
**Pull request by enricocolasante**

### CODE REVIEW

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionDaysBetween.java [hunk 1]
```diff
@@ -43,27 +42,15 @@ public static RuleFunctionDaysBetween create()
...
-                        Date startDate = format.parse( start );
-                        Date endDate = format.parse( end );
+                TimeInterval interval = getTimeInterval( start, end );
 
-                        return Long.valueOf( (endDate.getTime() - startDate.getTime()) / 86400000 ).intValue();
-                }
-                catch ( ParseException parseException )
-                {
-                        throw new RuntimeException( parseException );
+                if (interval.isEmpty()) {
```
Comment (Bekkalizer) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunctionDaysBetween.java]: Code style - missing spaces

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionDaysBetween.java [hunk 2]
```diff
@@ -43,27 +42,15 @@ public static RuleFunctionDaysBetween create()
...
+                        return 0;
                 }
+
+                return Days.daysBetween(interval.getStartDate(), interval.getEndDate()).getDays();
```
Comment (Bekkalizer) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunctionDaysBetween.java]: Code style, missing spaces

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionMonthsBetween.java [hunk 3]
```diff
@@ -60,56 +55,27 @@ public String evaluate(@Nonnull List<String> arguments, Map<String, RuleVariable
...
+     * Function which will return the number of months between the two given dates.
+     *
+     * @param start the start date.
+     * @param end   the end date.
+     * @return number of months between dates.
+     */
+    private Integer monthsBetween( String start, String end ) {
+        TimeInterval interval = getTimeInterval( start, end );
+
+        if (interval.isEmpty()) {
```
Comment (Bekkalizer) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunctionMonthsBetween.java]: Codestyle - spaces

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionMonthsBetween.java [hunk 4]
```diff
@@ -60,56 +55,27 @@ public String evaluate(@Nonnull List<String> arguments, Map<String, RuleVariable
...
+            return 0;
         }
 
-        long monthBetween = monthBetween(startDate, endDate);
-
-//        return String.valueOf(ChronoUnit.MONTHS.between(startDate, endDate));
-        return String.valueOf(monthBetween);
+        return Months.monthsBetween( interval.getStartDate(), interval.getEndDate()).getMonths();
```
Comment (Bekkalizer) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunctionMonthsBetween.java]: Codestyle - spaces

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionWeeksBetween.java [hunk 5]
```diff
@@ -30,6 +30,23 @@ public String evaluate( @Nonnull List<String> arguments,
...
+         * @return number of weeks between dates.
+         */
+        private Integer weeksBetween( String start, String end ) {
+                TimeInterval interval = getTimeInterval( start, end );
+
+                if (interval.isEmpty()) {
+                        return 0;
+                }
+
+                return Weeks.weeksBetween( interval.getStartDate(), interval.getEndDate()).getWeeks();
```
Comment (Bekkalizer) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunctionWeeksBetween.java]: Codestyle - spaces

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionWeeksBetween.java [hunk 6]
```diff
@@ -30,6 +30,23 @@ public String evaluate( @Nonnull List<String> arguments,
...
+
+        /**
+         * Function which will return the number of weeks between the two given dates.
+         *
+         * @param start the start date.
+         * @param end   the end date.
+         * @return number of weeks between dates.
+         */
+        private Integer weeksBetween( String start, String end ) {
+                TimeInterval interval = getTimeInterval( start, end );
```
Comment (Bekkalizer) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunctionWeeksBetween.java]: Codestyle - wrong indent

**Review by Bekkalizer — CHANGES_REQUESTED** [review]
Some code style, looks good otherwise!

**Review by Balcan — APPROVED** [review]
There are several unused imports in some classes. But apart from that, everything works well in android :)

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunction.java [hunk 7]
```diff
@@ -9,115 +12,128 @@
...
         }
+        catch ( final NumberFormatException nfe )
+        {
+            return defaultValue;
+        }
+    }
 
-        static boolean isEmpty( String input )
+    @Nonnull
+    public String wrap( String input )
```
Comment (Bekkalizer) [inline src/main/java/org/hisp/dhis/rules/functions/RuleFunction.java:125]: Perhaps trim single quotes, or check that we do not double wrap - if a user calls the function with hard coded dates with single quotes:
d2:daysBetween(#{event_Date},'2020-01-01')

Comment (enricocolasante) [reply src/main/java/org/hisp/dhis/rules/functions/RuleFunction.java:125]: I added a test to verify that single quoted dates are parsed in the correct way.
This wrap is used for output, and it is actually one of the changes that we need to verify in the Antlr PR

**Review by Bekkalizer — APPROVED** [review]

**Pull request merged**